### DISCUSSION TITLE
Permeability upscaling using grid bucket structure

### DISCUSSION THREAD

**Discussion by chrstmr** [pmgbergen/porepy#576] [Q&A]
Hello all, 
I am new to PorePy and am trying to upscale permeability after solving the single phase pressure equation on a simple box setting dirichlet pressure boundary conditions on two opposite sides. I want to include fractures into this box later, thus I am using a grid bucket structure for the box. gb = pp.meshing.cart_grid([], [3, 3, 3], physdims=[1, 1, 1])

I set the permeability to 1Darcy = 9.86923e-13 and pressure gradient dP to 1Pa (I did not find the conventions on units, but I assume SI units are used). The visualisation of the resulting pressure field after solving the system shows linear pressure distribution and looks fine to me. 

Now for the perm upscaling I need to solve the following equation k=q_in*L/A/dP, where q_in is the flux through the inlet boundary, L=1m length of the box, A=1m^2 the inlet area. Now the question is, how do I get the flux through the inlet boundary? 

I tried the following:
data = d['discretization_matrices']['flow']['flux']
q = np.sum(data[bool_inflow_faces])
k = q*L/A/dP

But what I am getting back is not k=1Darcy but 6Darcy. 
Any idea where my mistake could be? Is d['discretization_matrices']['flow']['flux'] the right place to get the boundary fluxes? 

Thank you for your help
Christine

**Comment by chrstmr** [comment]
Sometimes it helps to write things down. 
I figured out that the flux I am looking for is: 
d['discretization_matrices']['flow']['flux']*d['state']['pressure'] + \
                d['discretization_matrices']['flow']['bound_flux']*d['parameters']['flow']['bc_values']

or I can also use 
fvutils.compute_darcy_flux(g, data=d)
d['parameters']['flow']['darcy_flux']

please correct me if I am wrong 

Cheers, 
Christine

**Reply by keileg** [reply]
You're right. The first term, with `p` defines the parts of the flux determined by the interior degrees of freedom. The second term, with `bc_values` is r contribution from the boundary conditions. `compute_darcy_flux` is a convenience method which takes care of everything, this is particularly handy for problems with many fractures.

Does this make sense?

**Comment by chrstmr** [comment]
Thank you Eirik! 
While we are at it, what is the smartest way to save and load a gb object? I tried 

dump_grid_bucket_to_file(gb, 'gb_file_name')

but did not find any function to load the set of files that were created. 

Thanks 
Christine

**Comment by keileg** [comment]
Oh, that is an old function that was intended for other usage.

It's been a while since I saved a `GridBuket`, but I think the best option is to use the `pickle` module from the standard Python library - see for instance https://wiki.python.org/moin/UsingPickle for  an example. This should work - if not, I think we will consider it a bug, so feel free to report it.

One warning: If you save an object with pickle, update PorePy, and then try to load the object again, it may not work because pickle notes that the underlying code has changed in the meantime (ore something like that). A better way of storing the state of a problem have been on our wishlist for a long while, but so far, we have not found the time to do so.

**Reply by chrstmr** [reply]
Thank you for your reply, I will take a look at this library. Would be great if one could avoid the computational time of repeated meshing of the models.
